File: pegasusX/apps/payload-app-android/app/src/main/java/com/pegasus/payload/data/auth/FirebaseAuthHelper.kt
package com.pegasus.payload.data.auth

import android.app.Activity
import android.content.Context
import android.util.Log
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseException
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.PhoneAuthCredential
import com.google.firebase.auth.PhoneAuthOptions
import com.google.firebase.auth.PhoneAuthProvider
import com.pegasus.payload.BuildConfig
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.tasks.await
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Firebase Auth helper for dual-mode authentication.
 * Uses google-services.json; Auth emulator only when local.properties
 * sets firebase.auth.emulator=true.
 * All methods degrade gracefully — legacy JWT still works when Firebase is unavailable.
 */
object FirebaseAuthHelper {
    private const val TAG = "FirebaseAuth"
    private var initialized = false

    @Volatile
    private var pendingVerificationId: String? = null

    @Volatile
    private var autoCredential: PhoneAuthCredential? = null

    fun init(context: Context) {
        if (initialized) return
        try {
            if (FirebaseApp.getApps(context).isEmpty()) {
                FirebaseApp.initializeApp(context)
            }
            if (BuildConfig.FIREBASE_AUTH_EMULATOR) {
                FirebaseAuth.getInstance().useEmulator(BuildConfig.FIREBASE_AUTH_EMULATOR_HOST, 9099)
            }
            initialized = true
            Log.d(
                TAG,
                "Firebase Auth initialized (emulator=${BuildConfig.FIREBASE_AUTH_EMULATOR})",
            )
        } catch (e: Exception) {
            if (!BuildConfig.DEBUG) {
                throw IllegalStateException("Firebase configuration is missing in release build", e)
            }
            Log.w(TAG, "Firebase Auth init failed (non-fatal in dev): ${e.message}")
        }
    }

    suspend fun sendPhoneVerification(activity: Activity, phone: String): Unit =
        suspendCancellableCoroutine { cont ->
            val normalized = phone.trim()
            if (normalized.isBlank()) {
                cont.resumeWithException(IllegalArgumentException("Phone number required"))
                return@suspendCancellableCoroutine
            }
            val callbacks = object : PhoneAuthProvider.OnVerificationStateChangedCallbacks() {
                override fun onVerificationCompleted(credential: PhoneAuthCredential) {
                    pendingVerificationId = null
                    autoCredential = credential
                    if (cont.isActive) cont.resume(Unit)
                }

                override fun onVerificationFailed(e: FirebaseException) {
                    pendingVerificationId = null
                    Log.w(TAG, "Phone verification failed: ${e.message}")
                    if (cont.isActive) cont.resumeWithException(e)
                }

                override fun onCodeSent(verificationId: String, token: PhoneAuthProvider.ForceResendingToken) {
                    pendingVerificationId = verificationId
                    if (cont.isActive) cont.resume(Unit)
                }
            }
            val options = PhoneAuthOptions.newBuilder(FirebaseAuth.getInstance())
                .setPhoneNumber(normalized)
                .setTimeout(60L, TimeUnit.SECONDS)
                .setActivity(activity)
                .setCallbacks(callbacks)
                .build()
            PhoneAuthProvider.verifyPhoneNumber(options)
        }

    suspend fun verifySmsCode(code: String): String {
        val credential = autoCredential ?: run {
            val verificationId = pendingVerificationId
                ?: throw IllegalStateException("No verification in progress; request a code first")
            PhoneAuthProvider.getCredential(verificationId, code.trim())
        }
        autoCredential = null
        pendingVerificationId = null
        val result = FirebaseAuth.getInstance().signInWithCredential(credential).await()
        return result.user?.getIdToken(false)?.await()?.token
            ?: throw IllegalStateException("Firebase sign-in succeeded but id_token was missing")
    }

    fun hasAutoCredential(): Boolean = autoCredential != null

    suspend fun currentIdToken(): String? {
        return try {
            FirebaseAuth.getInstance().currentUser?.getIdToken(false)?.await()?.token
        } catch (e: Exception) {
            Log.w(TAG, "currentIdToken failed (non-fatal): ${e.message}")
            null
        }
    }

    suspend fun exchangeCustomToken(customToken: String): String? {
        if (customToken.isBlank()) return null
        return try {
            val result = FirebaseAuth.getInstance().signInWithCustomToken(customToken).await()
            result.user?.getIdToken(false)?.await()?.token
        } catch (e: Exception) {
            Log.w(TAG, "Custom token exchange failed (non-fatal): ${e.message}")
            null
        }
    }

    fun signOut() {
        pendingVerificationId = null
        autoCredential = null
        try {
            FirebaseAuth.getInstance().signOut()
        } catch (_: Exception) {
        }
    }
}

File: pegasusX/apps/payload-app-ios/payload-app-ios/Services/FirebaseAuthHelper.swift
//
//  FirebaseAuthHelper.swift
//  payload-app-ios
//
//  Firebase phone OTP + custom-token exchange for payload terminal auth.
//

import FirebaseAuth
import FirebaseCore
import Foundation

@MainActor
final class FirebaseAuthHelper {
    static let shared = FirebaseAuthHelper()

    private(set) var idToken: String?
    private var verificationID: String?
    private var initialized = false

    private init() {}

    func configure() {
        guard !initialized else { return }
        if FirebaseApp.app() == nil {
            let options = FirebaseOptions(
                googleAppID: "1:000000000000:ios:0000000000000001",
                gcmSenderID: "000000000000"
            )
            options.projectID = "demo-pegasus"
            options.apiKey = "demo-key"
            FirebaseApp.configure(options: options)
        }
        #if DEBUG
        Auth.auth().useEmulator(withHost: "localhost", port: 9099)
        #endif
        initialized = true
    }

    func sendPhoneVerification(phone: String) async throws {
        configure()
        let trimmed = phone.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw NSError(domain: "FirebaseAuth", code: 1, userInfo: [NSLocalizedDescriptionKey: "Phone number required"])
        }
        let verificationID = try await withCheckedThrowingContinuation { (cont: CheckedContinuation<String, Error>) in
            PhoneAuthProvider.provider().verifyPhoneNumber(trimmed, uiDelegate: nil) { verificationID, error in
                if let error {
                    cont.resume(throwing: error)
                    return
                }
                guard let verificationID else {
                    cont.resume(throwing: NSError(
                        domain: "FirebaseAuth",
                        code: 2,
                        userInfo: [NSLocalizedDescriptionKey: "Missing verification ID"]
                    ))
                    return
                }
                cont.resume(returning: verificationID)
            }
        }
        self.verificationID = verificationID
    }

    func verifySmsCode(_ code: String) async throws -> String {
        configure()
        guard let verificationID else {
            throw NSError(
                domain: "FirebaseAuth",
                code: 3,
                userInfo: [NSLocalizedDescriptionKey: "No verification in progress; request a code first"]
            )
        }
        let credential = PhoneAuthProvider.provider().credential(
            withVerificationID: verificationID,
            verificationCode: code.trimmingCharacters(in: .whitespacesAndNewlines)
        )
        let result = try await Auth.auth().signIn(with: credential)
        self.verificationID = nil
        let token = try await result.user.getIDToken()
        idToken = token
        return token
    }

    func exchangeCustomToken(_ customToken: String, completion: @escaping (String?) -> Void) {
        configure()
        guard !customToken.isEmpty else {
            completion(nil)
            return
        }
        Auth.auth().signIn(withCustomToken: customToken) { result, error in
            if let error {
                print("[FirebaseAuth] Custom token exchange failed: \(error.localizedDescription)")
                completion(nil)
                return
            }
            result?.user.getIDToken { token, _ in
                self.idToken = token
                completion(token)
            }
        }
    }

    func signOut() {
        verificationID = nil
        idToken = nil
        try? Auth.auth().signOut()
    }
}
